Changes
=======

Version 1.1.1:
     * Change to options passed to **cutadapt** : add the '-m 4' option to discard any reads which are less than 4bp after adapters have been trimmed. Without this fastq files could contain zero length reads which resulted in an error being generated by **bowtie**. *(In future could try to keep the mate of such reads, but currently both members of the pair are discarded.)*

     * Set default exclusion zone to 1000bp. Previously this was 500bp, but since the DpnII fragments are on average ~300bp, and the exclusion zone is measured by taking the fragment centre-to-centre distance, probably this was quite short for most experiments.
 
Version 1.1.0:
     * Added 'downstream/' directory which contains several 'worksheets' (in html/pdf/Rmd format) showing examples of downstream analysis of capC-MAP output.
     * Added some python functions to the library in capC/capCtools. This includes a function ``report()`` which reads a capC-MAP report file, returning a python dictionary containing various statistics from the report, and a function ``read_interactioncounts()`` which reads the interaction counts file generated by capC-MAP and returns a dictionary with 'per target' statitics (such as inter vs interchromosomal reads etc.).
     * Added a feature such that capC-MAP now counts the number of local (<1Mbp or <5Mb) interactions for each target, outputting these alongside per target counts of intra and inter chromosomal interactions.
     * Fixed a bug where file paths inducing ``~/`` were not always properly expanded.

